### PR TITLE
WIP Use lzfse as a portable alternative to libcompression.

### PR DESCRIPTION
**Pull request by grp**
WIP pending https://github.com/lzfse/lzfse/pull/25 and https://github.com/lzfse/lzfse/pull/26.

Allows decompressing LZFSE and LZVN compression on Linux.

Should the lzfse submodule be required? That would make the default build require git submodules be cloned, which isn't necessarily right now. Alternatively, could make either libcompression or lzfse required, or leave them both as optional.

### CODE REVIEW

**Comment by facebook-github-bot-9** [comment]
Thank you for your pull request.  As you may know, we require contributors to sign our Contributor License Agreement, and we don't seem to have you on file and listed as active anymore.  In order for us to review and merge your code, please email cla@fb.com with your details so we can update your status.

**Comment by facebook-github-bot-8** [comment]
Thank you for signing our Contributor License Agreement. We can now accept your code for this (and any) Facebook open source project. Thanks!

**Comment by marcsalem** [comment]
I would be in favor of a submodule in this case since the library is much less mature (changing faster than libpng) and less common to see it supported already in a distro.

But use HTTPS not SSH.

File: Libraries/libcar/CMakeLists.txt [hunk 1]
```diff
@@ -20,12 +20,24 @@ find_package(ZLIB REQUIRED)
...
+  set(LZFSE_BINARY_DIR "${PROJECT_BINARY_DIR}/lzfse") 
+  set(LZFSE_BUNDLE_MODE FALSE)
+  add_subdirectory("${LZFSE_SOURCE_DIR}" "${LZFSE_BINARY_DIR}" EXCLUDE_FROM_ALL)
+
+  target_compile_definitions(car PRIVATE HAVE_LZFSE=1)
+  target_link_libraries(car PRIVATE lzfse)
+  target_include_directories(car PRIVATE "${LZFSE_SOURCE_DIR}/src")
+endif ()
+
 find_library(COMPRESSION compression)
```
Comment (strager) [inline Libraries/libcar/CMakeLists.txt]: Don't bother searching for compression if LZFSE is found.

**Comment by strager** [comment]
> Should the lzfse submodule be required? That would make the default build require git submodules be cloned, which isn't necessarily right now. Alternatively, could make either libcompression or lzfse required, or leave them both as optional.

The "CMake way" is to:

* Expose an option called `XCBUILD_USE_LZFSE` (or something) to enable/disable lzfse support,
* Create a CMake module called `FindLzfse.cmake`,
* If `XCBUILD_USE_LZFSE` is truthy:
 * Use your module: `find_package(Lzfse)`,
 * Fall back to compiling the source version if `LZFSE_FOUND` is falsy and the source version exists, and
 * Error out if `XCBUILD_USE_LZFSE` was explicitly set and `LZFSE_FOUND` is falsy.

(Things may be a bit different lzfse wants to use pkgconfig or CPack.)

This gives the user the power to configure things how they want.

* If the user doesn't care about lzfse support, they don't need to do anything. Things will configure whether or not lzfse is installed or its source is checked out.
* If lzfse ever becomes stable, you don't need to touch xcbuild's CMakeLists.txt in order for users to benefit from their system lzfse.
* If the system lzfse is incompatible with xcbuild, `-DLZFSE_FOUND=YES` falls back to the in-source version.
* If the user has a custom lzfse build, they can configure flags with `-DLZFSE_LIBRARIES=/path/to/lzfse.a` and such. (This is how Facebook would configure xcbuild.)
* If the user doesn't want lzfse support, `-DXCBUILD_USE_LZFSE=NO` disables everything.

**Comment by strager** [comment]
TL;DR provide a way for Facebook to configure where lzfse's libs/includes or project source is.

**Comment by grp** [comment]
@strager Sounds good. Is it OK for Facebook if I make this require my fork of lzfse (https://github.com/grp/lzfse) instead, to access the LZVN APIs? Or would it be preferable to reach into the lzfse internals from upstream inside xcbuild?

**Comment by strager** [comment]
> Is it OK for Facebook if I make this require https://github.com/grp/lzfse instead, to access the LZVN APIs?

Yeah, that's fine.
